lib/Auth/Process/PerunEntitlementExtended.php

```diff
@@ -0,0 +1,148 @@
+
+        $this->outputAttrName = $configuration->getString(self::OUTPUT_ATTR_NAME, 'eduPersonEntitlementExtended');
+        $this->releaseForwardedEntitlement = $configuration->getBoolean(self::RELEASE_FORWARDED_ENTITLEMENT, true);
+        $this->forwardedEduPersonEntitlement = $configuration->getString(
+            self::FORWARDED_EDU_PERSON_ENTITLEMENT,
+            $this->releaseForwardedEntitlement ? Configuration::REQUIRED_OPTION : ''
+        );
+
+        $this->groupNameAARC = $modulePerunConfiguration->getBoolean(self::GROUPNAMEAARC_ATTR, false);
+        $this->entitlementPrefix = $modulePerunConfiguration->getString(
+            self::ENTITLEMENTPREFIX_ATTR,
+            $this->groupNameAARC ? Configuration::REQUIRED_OPTION : ''
+        );
+        $this->entitlementAuthority = $modulePerunConfiguration->getString(
+            self::ENTITLEMENTAUTHORITY_ATTR,
+            $this->groupNameAARC ? Configuration::REQUIRED_OPTION : ''
+        );
+
+        $interface = $configuration->getValueValidate(
+            self::INTERFACE_PROPNAME,
+            [Adapter::RPC, Adapter::LDAP],
+            Adapter::RPC
+        );
+        $this->adapter = Adapter::getInstance($interface);
+    }
+
+    public function process(&$request)
+    {
+        $eduPersonEntitlementExtended = [];
+        $capabilities = [];
+        $forwardedEduPersonEntitlement = [];
+
+        if (isset($request['perun']['groups'])) {
+            $eduPersonEntitlementExtended = $this->getEduPersonEntitlementExtended($request);
+
+            $capabilities = EntitlementUtils::getCapabilities(
+                $request,
+                $this->adapter,
+                $this->entitlementPrefix,
+                $this->entitlementAuthority
+            );
+        } else {
+            Logger::debug(
+                'perun:PerunEntitlementExtended: There are no user groups assigned to facility.' .
+                '=> Skipping getEduPersonEntitlementExtended and getCapabilities'
+            );
+        }
+
+        if ($this->releaseForwardedEntitlement) {
+            $forwardedEduPersonEntitlement = EntitlementUtils::getForwardedEduPersonEntitlement(
+                $request,
+                $this->adapter,
+                $this->forwardedEduPersonEntitlement
+            );
+        }
+
+        $request['Attributes'][$this->outputAttrName] = array_unique(array_merge(
+            $eduPersonEntitlementExtended,
+            $forwardedEduPersonEntitlement,
+            $capabilities
+        ));
+    }
+
+    private function getEduPersonEntitlementExtended(&$request)
+    {
+        $eduPersonEntitlementExtended = [];
+
+        $groups = $request['perun']['groups'];
+        foreach ($groups as $group) {
+            array_push($eduPersonEntitlementExtended, $this->groupNameWrapper($group->getUuid()));
+
+            $groupName = $group->getUniqueName();
+            $groupName = preg_replace('/^(\w*)\:members$/', '$1', $groupName);
```

Comment:
I would move these two lines either into the "wrapper" or closer to line 124 (as it uses it as a passed parameter).